## **Exam Scores Prediction**

The dataset contains test scores from three different subjects at a fictional public school. This dataset also contains variables about various personal and socio-economic factors for each student.

The goal is to provide an analysis on how several variables could affect a student's exam score and to predict future exam scores given those variables.

*Credits to the creator of the original dataset, **Mr. Royce Kimmons**, and to the uploaded dataset in **[Kaggle](https://www.kaggle.com/datasets/desalegngeb/students-exam-scores/data)***.

### **Data Description**

Column | Description |
-----|-----|
Gender | Gender of the student (male/female) |
EthnicGroup | Ethnic group of the student (group A to E) |
ParentEduc | Parent(s) education background (from some_highschool to master's degree) |
LunchType | School lunch type (standard or free/reduced) |
TestPrep | Test preparation course followed (completed or none) |
ParentMaritalStatus | Parent(s) marital status (married/single/widowed/divorced) |
PracticeSport | How often the student practice sport (never/sometimes/regularly) |
IsFirstChild | If the first child in the family or not (yes/no) |
NrSiblings | Number of siblings the student has (0 to 7) |
TransportMeans | Means of transport to school (schoolbus/private) |
WklyStudyHours | Weekly self-study hours(less than 5hrs, between 5 and 10 hrs, more than 10 hours) |
MathScore | Math test score (0-100) |
ReadingScore | Reading test score (0-100) |
WritingScore | Writing test score (0-100) |

### **Preparation**

In [174]:
## For data manipulation

import numpy as np
import pandas as pd

## For data visualization

import matplotlib.pyplot as plt
import seaborn as sns

## For displaying all columns in the dataframe

pd.set_option('display.max_columns', None)

## For statistical tests

from scipy import stats

## For data modelling

from xgboost import XGBClassifier
from xgboost import XGBRegressor
from xgboost import plot_importance
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

## For metrics and helpful functions

from sklearn.model_selection import GridSearchCV, train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score, confusion_matrix, ConfusionMatrixDisplay, classification_report, roc_curve
from sklearn.tree import plot_tree

## For saving models

import pickle

## Miscellaneous

import warnings
warnings.simplefilter(action='ignore', category=FutureWarning)

### **Data Exploration**

In [175]:
## Load the dataset

df = pd.read_csv("D:\Documents\CSV Datasets\Student Marks\Expanded_data_with_more_features.csv")

df.head(10)

Unnamed: 0  Gender EthnicGroup          ParentEduc     LunchType  \
0           0  female         NaN   bachelor's degree      standard   
1           1  female     group C        some college      standard   
2           2  female     group B     master's degree      standard   
3           3    male     group A  associate's degree  free/reduced   
4           4    male     group C        some college      standard   
5           5  female     group B  associate's degree      standard   
6           6  female     group B        some college      standard   
7           7    male     group B        some college  free/reduced   
8           8    male     group D         high school  free/reduced   
9           9  female     group B         high school  free/reduced   

    TestPrep ParentMaritalStatus PracticeSport IsFirstChild  NrSiblings  \
0       none             married     regularly          yes         3.0   
1        NaN             married     sometimes          yes         0.0   
2       none              single     sometimes          yes         4.0   
3       none             married         never           no         1.0   
4       none             married     sometimes          yes         0.0   
5       none             married     regularly          yes         1.0   
6  completed             widowed         never           no         1.0   
7       none             married     sometimes          yes         1.0   
8  completed              single     sometimes           no         3.0   
9       none             married     regularly          yes         NaN   

  TransportMeans WklyStudyHours  MathScore  ReadingScore  WritingScore  
0     school_bus            < 5         71            71            74  
1            NaN         5 - 10         69            90            88  
2     school_bus            < 5         87            93            91  
3            NaN         5 - 10         45            56            42  
4     school_bus         5 - 10         76            78            75  
5     school_bus         5 - 10         73            84            79  
6        private         5 - 10         85            93            89  
7        private           > 10         41            43            39  
8        private           > 10         65            64            68  
9        private            < 5         37            59            50

In [176]:
## Check the size of the dataset

print(df.shape)

(30641, 15)


The dataset contains **`30,641`** rows and **`15`** columns.

In [177]:
## Check information about the dataset

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30641 entries, 0 to 30640
Data columns (total 15 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Unnamed: 0           30641 non-null  int64  
 1   Gender               30641 non-null  object 
 2   EthnicGroup          28801 non-null  object 
 3   ParentEduc           28796 non-null  object 
 4   LunchType            30641 non-null  object 
 5   TestPrep             28811 non-null  object 
 6   ParentMaritalStatus  29451 non-null  object 
 7   PracticeSport        30010 non-null  object 
 8   IsFirstChild         29737 non-null  object 
 9   NrSiblings           29069 non-null  float64
 10  TransportMeans       27507 non-null  object 
 11  WklyStudyHours       29686 non-null  object 
 12  MathScore            30641 non-null  int64  
 13  ReadingScore         30641 non-null  int64  
 14  WritingScore         30641 non-null  int64  
dtypes: float64(1), int64(4), object(10)


#### **Check for *null* values**

In [178]:
## Check the number of null values for each variable in the dataset

for col in df.columns:
    print('Null Values for column {} is {}%'.format(col, np.round(df[col].isnull().sum()*100 / len(df[col])),2))

Null Values for column Unnamed: 0 is 0.0%
Null Values for column Gender is 0.0%
Null Values for column EthnicGroup is 6.0%
Null Values for column ParentEduc is 6.0%
Null Values for column LunchType is 0.0%
Null Values for column TestPrep is 6.0%
Null Values for column ParentMaritalStatus is 4.0%
Null Values for column PracticeSport is 2.0%
Null Values for column IsFirstChild is 3.0%
Null Values for column NrSiblings is 5.0%
Null Values for column TransportMeans is 10.0%
Null Values for column WklyStudyHours is 3.0%
Null Values for column MathScore is 0.0%
Null Values for column ReadingScore is 0.0%
Null Values for column WritingScore is 0.0%


Some columns have ***null*** values. However, we will check the best method to handle those missing values.

Also, the Dtype for the column **NrSiblings** is ***float64***. This is not the correct datatype because the column contains *discrete* variables. The datatype for this column should be changed to ***int64***.

#### **Converting datatypes, renaming columns, and removing unnecessary columns**

In [179]:
## Changing the datatype of 'NrSiblings' into int64

df['NrSiblings'] = pd.to_numeric(df['NrSiblings'], downcast='integer', errors='coerce')
df['NrSiblings'] = df['NrSiblings'].astype('Int64')

df['NrSiblings'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 30641 entries, 0 to 30640
Series name: NrSiblings
Non-Null Count  Dtype
--------------  -----
29069 non-null  Int64
dtypes: Int64(1)
memory usage: 269.4 KB


In [180]:
## Removed the 'Unnamed' column

df = df.drop('Unnamed: 0', axis=1)

In [181]:
## Change column names to 'snake_case'

df.rename(columns={'Gender':'gender',
                   'EthnicGroup':'ethnic_group',
                   'ParentEduc':'parents_education',
                   'LunchType':'lunch_type',
                   'TestPrep':'test_preparation',
                   'ParentMaritalStatus':'parent_marital_status',
                   'PracticeSport':'practice_sports',
                   'IsFirstChild':'is_first_child',
                   'NrSiblings':'number_of_siblings',
                   'TransportMeans':'transport_means',
                   'WklyStudyHours':'weekly_study_hours',
                   'MathScore':'math_score',
                   'ReadingScore':'reading_score',
                   'WritingScore':'writing_score'}, inplace=True)

df.head()

gender ethnic_group   parents_education    lunch_type test_preparation  \
0  female          NaN   bachelor's degree      standard             none   
1  female      group C        some college      standard              NaN   
2  female      group B     master's degree      standard             none   
3    male      group A  associate's degree  free/reduced             none   
4    male      group C        some college      standard             none   

  parent_marital_status practice_sports is_first_child  number_of_siblings  \
0               married       regularly            yes                   3   
1               married       sometimes            yes                   0   
2                single       sometimes            yes                   4   
3               married           never             no                   1   
4               married       sometimes            yes                   0   

  transport_means weekly_study_hours  math_score  reading_score  writing_score  
0      school_bus                < 5          71             71             74  
1             NaN             5 - 10          69             90             88  
2      school_bus                < 5          87             93             91  
3             NaN             5 - 10          45             56             42  
4      school_bus             5 - 10          76             78             75

### **Handling Missing Data**

The method for handling missing data will be either by ***removing the null values***, or by ***imputation***.

#### **Method 1A: Removing *null* values in the dataset**

In [182]:
## Drop null values from all columns

df1 = df.dropna(axis=0)

df1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 19243 entries, 2 to 30640
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   gender                 19243 non-null  object
 1   ethnic_group           19243 non-null  object
 2   parents_education      19243 non-null  object
 3   lunch_type             19243 non-null  object
 4   test_preparation       19243 non-null  object
 5   parent_marital_status  19243 non-null  object
 6   practice_sports        19243 non-null  object
 7   is_first_child         19243 non-null  object
 8   number_of_siblings     19243 non-null  Int64 
 9   transport_means        19243 non-null  object
 10  weekly_study_hours     19243 non-null  object
 11  math_score             19243 non-null  int64 
 12  reading_score          19243 non-null  int64 
 13  writing_score          19243 non-null  int64 
dtypes: Int64(1), int64(3), object(10)
memory usage: 2.2+ MB


After removing all the *null* values in the dataset, we are left with **19,243** rows, which is 62.80% of the original dataset.

#### **Method 1B: Removing *null* values from specified columns**

To avoid over-representation of certain data, the limit for the number of *null* values in which imputation will be used will be set to **5.0%** of the total number of data in a certain column.

The columns **EthnicGroup**, **ParentEduc**, **TestPrep**, and **TransportMeans** contains *null* values that are more than 5.0% of the total number of data in their respective column. We will remove *null* values from this columns before proceeding with imputation.

In [183]:
## Drop null values from selected columns

df1b = df.copy()

df1b = df1b.dropna(subset=['ethnic_group','parents_education','test_preparation','number_of_siblings','transport_means'], axis=0)

df1b.info()

<class 'pandas.core.frame.DataFrame'>
Index: 21721 entries, 2 to 30640
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   gender                 21721 non-null  object
 1   ethnic_group           21721 non-null  object
 2   parents_education      21721 non-null  object
 3   lunch_type             21721 non-null  object
 4   test_preparation       21721 non-null  object
 5   parent_marital_status  20883 non-null  object
 6   practice_sports        21298 non-null  object
 7   is_first_child         21094 non-null  object
 8   number_of_siblings     21721 non-null  Int64 
 9   transport_means        21721 non-null  object
 10  weekly_study_hours     21038 non-null  object
 11  math_score             21721 non-null  int64 
 12  reading_score          21721 non-null  int64 
 13  writing_score          21721 non-null  int64 
dtypes: Int64(1), int64(3), object(10)
memory usage: 2.5+ MB


#### **Method 2: Use *Mode Imputation***

The values for the columns with *null* data are contains categorical and binary data. To avoid over-representing certain data, we set a limit to null values equal to or less than **5.0%** of the total number of data in their respective column. The columns with null values that exceeded the limit had all their null values dropped.

In [185]:
## Mode imputation for categorical columns

df1b['parent_marital_status'] = df1b['parent_marital_status'].fillna(df1b['parent_marital_status'].mode()[0])
df1b['practice_sports'] = df1b['practice_sports'].fillna(df1b['practice_sports'].mode()[0])
df1b['weekly_study_hours'] = df1b['weekly_study_hours'].fillna(df1b['weekly_study_hours'].mode()[0])

## Mode imputation for binary columns

df1b['is_first_child'] = df1b['is_first_child'].fillna(df1b['is_first_child'].mode()[0])

df1b.isnull().sum()

gender                   0
ethnic_group             0
parents_education        0
lunch_type               0
test_preparation         0
parent_marital_status    0
practice_sports          0
is_first_child           0
number_of_siblings       0
transport_means          0
weekly_study_hours       0
math_score               0
reading_score            0
writing_score            0
dtype: int64

#### **Compare the descriptive statistics of the two datasets**

In [186]:
descriptive_stats = pd.concat([df.describe(), df1b.describe()], axis=1, keys=['Dataset with Removed Null Values','Dataset with Mode Imputation'])

descriptive_stats

Dataset with Removed Null Values                              \
                    number_of_siblings    math_score reading_score   
count                          29069.0  30641.000000  30641.000000   
mean                          2.145894     66.558402     69.377533   
std                           1.458242     15.361616     14.758952   
min                                0.0      0.000000     10.000000   
25%                                1.0     56.000000     59.000000   
50%                                2.0     67.000000     70.000000   
75%                                3.0     78.000000     80.000000   
max                                7.0    100.000000    100.000000   

                    Dataset with Mode Imputation                              \
      writing_score           number_of_siblings    math_score reading_score   
count  30641.000000                      21721.0  21721.000000  21721.000000   
mean      68.418622                     2.140785     66.589844     69.467152   
std       15.443525                     1.447413     15.382437     14.787346   
min        4.000000                          0.0      0.000000     10.000000   
25%       58.000000                          1.0     56.000000     59.000000   
50%       69.000000                          2.0     67.000000     70.000000   
75%       79.000000                          3.0     78.000000     80.000000   
max      100.000000                          7.0    100.000000    100.000000   

                     
      writing_score  
count  21721.000000  
mean      68.534736  
std       15.488860  
min        4.000000  
25%       58.000000  
50%       69.000000  
75%       79.000000  
max      100.000000

Based on the descriptive statistics from the resulting datasets from the two methods, the statistic values for the dataset with mode imputations applied varies from the dataset with all null values removed by a very small amount (less than 1.00).

With this result, we will proceed on using the dataset obtained from mode imputation as it contains more data compared to the dataset with all null values removed.